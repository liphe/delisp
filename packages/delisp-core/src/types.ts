//
// Types
//

interface TConstantF<_T> {
  tag: "constant";
  name: string;
}

interface TApplicationF<T> {
  tag: "application";
  op: T;
  args: T[];
}

interface TVarF<_T> {
  tag: "type-variable";
  name: string;
  userSpecified: boolean;
}

interface REmptyF<_T> {
  tag: "empty-row";
}

interface RExtensionF<T> {
  tag: "row-extension";
  label: string;
  labelType: T;
  extends: T;
}

type RowF<T> = REmptyF<T> | RExtensionF<T>;
type AnyTypeF<T = Type> = TConstantF<T> | TApplicationF<T> | TVarF<T> | RowF<T>;

interface Node<A> {
  node: A;
}

export interface TConstant extends Node<TConstantF<Type>> {}
export interface TApplication extends Node<TApplicationF<Type>> {}
export interface TVar extends Node<TVarF<Type>> {}
export interface REmpty extends Node<REmptyF<Type>> {}
export interface RExtension extends Node<RExtensionF<Type>> {}
export interface Row extends Node<RowF<Type>> {}

export interface Type extends Node<AnyTypeF<Type>> {}
export type TypeF<A> = Node<AnyTypeF<A>>;

export interface TypeSchema {
  tag: "type";
  tvars: string[];
  mono: Type;
}

//
// Constructor helpers
//

function tConstant(name: string): TConstant {
  return { node: { tag: "constant", name } };
}

// * ... -> effect -> *
export const tcArrow = tConstant("->");

// * ... -> *
export const tcStar = tConstant("*");

// * -> *
export const tcVector = tConstant("vector");
// row -> *
export const tcRecord = tConstant("record");
// row -> *
export const tcCases = tConstant("cases");
// *
export const tVoid = tConstant("void");
export const tBoolean = tConstant("boolean");
export const tNumber = tConstant("number");
export const tString = tConstant("string");
export const tNone = tConstant("none");

// row -> effect
const tcEffect = tConstant("effect");
// row -> values
const tcValues = tConstant("values");

export function tVar(name: string, userSpecified = false): TVar {
  return {
    node: {
      tag: "type-variable",
      name,
      userSpecified
    }
  };
}

export function tUserDefined(name: string): TConstant {
  return { node: { tag: "constant", name } };
}

export function tApp(op: Type, ...args: Type[]): Type {
  return {
    node: {
      tag: "application",
      op,
      args
    }
  };
}

export function tVector(t: Type): Type {
  return tApp(tcVector, t);
}

export const emptyRow: REmpty = {
  node: { tag: "empty-row" }
};

export const tRowExtension = (
  label: string,
  labelType: Type,
  row: Type
): RExtension => ({
  node: {
    tag: "row-extension",
    label,
    labelType,
    extends: row
  }
});

export function tRow(
  fields: Array<{ label: string; type: Type }>,
  extending: Type = emptyRow
): Type {
  return fields.reduceRight(
    (row: Type, { label, type }): Row => tRowExtension(label, type, row),
    extending
  );
}

export function tEffect(labels: string[], extending: Type = emptyRow): Type {
  return tApp(
    tcEffect,
    tRow(labels.map(label => ({ label, type: tVoid })), extending)
  );
}

export function tRecord(
  fields: Array<{ label: string; type: Type }>,
  extending: Type = emptyRow
): Type {
  return tApp(tcRecord, tRow(fields, extending));
}

export function tCases(
  variants: Array<{ label: string; type: Type }>,
  extending: Type = emptyRow
): Type {
  return tApp(tcCases, tRow(variants, extending));
}

export function tProduct(types: Type[]) {
  return tApp(tcStar, ...types);
}

export function tValues(types: Type[], extending: Type = emptyRow): Type {
  return tApp(
    tcValues,
    tRow(types.map((type, i) => ({ label: String(i), type })), extending)
  );
}

export function tMultiValuedFunction(
  args: Type[],
  effect: Type,
  out: Type
): Type {
  return tApp(tcArrow, ...args, effect, out);
}

export function tFn(args: Type[], effect: Type, out: Type): Type {
  return tMultiValuedFunction(args, effect, tValues([out]));
}
