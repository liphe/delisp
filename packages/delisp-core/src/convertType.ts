import { printHighlightedExpr } from "./error-report";
import { ASExpr, ASExprList, ASExprSymbol } from "./sexpr";
import { Type } from "./types";

function convertSymbol(expr: ASExprSymbol): Type {
  switch (expr.name) {
    case "number":
      return { type: "number" };
    case "string":
      return { type: "string" };
    default:
      return { type: "type-variable", name: expr.name };
  }
}

function convertList(expr: ASExprList): Type {
  const [first, from, to] = expr.elements;

  if (first.type !== "symbol" || first.name !== "->") {
    throw new Error(printHighlightedExpr("Expected -> as first symbol", expr));
  }

  if (expr.elements.length !== 3) {
    throw new Error(printHighlightedExpr("Expected 3 elements", expr));
  }

  if (from.type !== "list") {
    throw new Error(
      printHighlightedExpr("Expected second element to be a list", expr)
    );
  }

  return {
    type: "function",
    from: from.elements.map(convert),
    to: convert(to)
  };
}

export function convert(expr: ASExpr): Type {
  switch (expr.type) {
    case "list":
      return convertList(expr);
    case "symbol":
      return convertSymbol(expr);
    default:
      throw new Error(printHighlightedExpr("Not a valid type", expr));
  }
}
