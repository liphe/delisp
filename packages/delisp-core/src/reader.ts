//
// reader.ts --- The reader for Deisp
//

//
// S-Expression parser
//

import {
  alternatives,
  atLeastOne,
  character,
  delimited,
  delimitedMany,
  endOfInput,
  many,
  Parser,
  until
} from "./parser-combinators";

import { ASExpr } from "./sexpr";

const spaces = many(
  alternatives(character(" "), character("\t"), character("\n"))
);

function spaced<A>(x: Parser<A>) {
  return delimited(spaces, x, spaces);
}

const alphanumeric = character().chain(char => {
  if (/[a-zA-Z0-9]/.test(char)) {
    return Parser.of(char);
  } else {
    return Parser.fail("an alphanumeric character");
  }
});

const digit = character().chain(char => {
  if (/[0-9]/.test(char)) {
    return Parser.of(char);
  } else {
    return Parser.fail("a digit");
  }
});

//
// Number
//

const unsignedInteger: Parser<number> = atLeastOne(digit).map(digits =>
  parseInt(digits.join(""), 10)
);
const negativeInteger: Parser<number> = character("-")
  .chain(_ => unsignedInteger)
  .map(num => -num);

const numberP: Parser<ASExpr> = alternatives(negativeInteger, unsignedInteger)
  .map(
    (val, location): ASExpr => ({
      type: "number",
      value: val,
      location
    })
  )
  .description("number");

//
// String
//
const doubleQuote = character('"').description("double quote");

const stringChar = character()
  .chain(char => {
    if (char === '"') {
      return Parser.fail("Not inside string");
    }
    if (char === "\\") {
      return alternatives(
        character("n").map(_ => "\n"),
        character("\\")
      ).description("escaped string character");
    }
    return Parser.of(char);
  })
  .description("string character");

const stringP = delimitedMany(doubleQuote, stringChar, doubleQuote)
  .map(
    (chars, location): ASExpr => ({
      type: "string",
      value: chars.join(""),
      location
    })
  )
  .description("string");

//
// Symbol
//

const symbol: Parser<ASExpr> = atLeastOne(
  alternatives(
    alphanumeric,
    character("!"),
    character("@"),
    character("#"),
    character("$"),
    character("%"),
    character("^"),
    character("&"),
    character("*"),
    character("<"),
    character(">"),
    character("+"),
    character("-"),
    character("/"),
    character("~"),
    character("?")
  )
)
  .map(
    (chars, location): ASExpr => ({
      type: "symbol",
      name: chars.join(""),
      location
    })
  )
  .description("symbol");

//
// Lists & S-Expressions
//

const atom: Parser<ASExpr> = alternatives(numberP, stringP, symbol).description(
  "atom"
);

const leftParen = character("(").description("open parenthesis");
const rightParen = character(")").description("close parenthesis");

function list(x: Parser<ASExpr>): Parser<ASExpr> {
  return delimitedMany(leftParen, x, spaces.then(rightParen))
    .map(
      (elements, location): ASExpr => ({
        type: "list",
        elements,
        location
      })
    )
    .description("list");
}

const reportUnmatched: Parser<{}> = Parser.lookahead(rightParen).chain(
  closed => {
    if (closed) {
      return Parser.fail("Unmatched closed parenthesis");
    } else {
      return Parser.of({});
    }
  }
);

const sexpr: Parser<ASExpr> = spaced(
  reportUnmatched.then(atom.or(() => list(sexpr)))
);

const sexprs: Parser<ASExpr[]> = until(endOfInput, sexpr);

export function readAllFromString(str: string): ASExpr[] {
  return sexprs.parse(str);
}

//
// Parser a Delisp expression from a string
//
export function readFromString(str: string): ASExpr {
  return sexpr.parse(str);
}
