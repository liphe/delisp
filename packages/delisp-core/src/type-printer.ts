import { InvariantViolation } from "./invariant";
import { range } from "./utils";
import { TApplication, Type, TVar } from "./types";
import { normalizeRow } from "./type-utils";

function typeIndexName(index: number): string {
  const alphabet = "αβγδεζηθικμνξοπρστυφχψ";
  return index < alphabet.length
    ? alphabet[index]
    : `ω${index - alphabet.length + 1}`;
}

function createVariableNormalizer() {
  let variables: string[] = [];

  return (t: TVar): string => {
    let idx = variables.indexOf(t.node.name);
    if (idx === -1) {
      idx = variables.length;
      variables.push(t.node.name);
    }
    return typeIndexName(idx);
  };
}

function printValuesType(
  fnout: Type,
  normalizeVar: (t: TVar) => string,
  simplify = false
): string {
  const row = normalizeRow(fnout);

  // primary value only
  if (simplify && row.fields.length === 1 && row.fields[0].label === "0") {
    return _printType(row.fields[0].labelType, normalizeVar);
  } else {
    // full form (values a b c ...)
    return (
      "(values " +
      range(row.fields.length)
        .map(i => {
          const field = row.fields.find(f => f.label === String(i))!;
          return _printType(field.labelType, normalizeVar);
        })
        .join(" ") +
      (row.extends.node.tag === "empty-row"
        ? ""
        : "| " + _printType(row.extends, normalizeVar)) +
      ")"
    );
  }
}

function printApplicationType(
  type: TApplication,
  normalizeVar: (t: TVar) => string,
  simplify = false
): string {
  function genericApp(op: Type, args: Type[]): string {
    return (
      "(" + [op, ...args].map(t => _printType(t, normalizeVar)).join(" ") + ")"
    );
  }

  if (type.node.op.node.tag === "constant") {
    switch (type.node.op.node.name) {
      case "vector": {
        return `[${_printType(type.node.args[0], normalizeVar)}]`;
      }
      case "record": {
        const arg = type.node.args[0];
        const row = normalizeRow(arg);
        const fields = row.fields
          .map(f => `${f.label} ${_printType(f.labelType, normalizeVar)}`)
          .join(" ");
        const extension =
          row.extends.node.tag !== "empty-row"
            ? ` | ${_printType(row.extends, normalizeVar)}`
            : "";
        return `{${fields}${extension}}`;
      }
      case "effect": {
        const row = normalizeRow(type.node.args[0]);

        const fields = row.fields.map(f => " " + f.label).join("");
        const extending =
          row.extends.node.tag === "empty-row"
            ? ""
            : " | " + _printType(row.extends, normalizeVar);

        return `(effect${fields}${extending})`;
      }

      case "cases": {
        const arg = type.node.args[0];
        const row = normalizeRow(arg);
        const fields = row.fields
          .map(f =>
            f.labelType.node.tag === "constant" &&
            f.labelType.node.name === "void"
              ? `${f.label}`
              : `(${f.label} ${_printType(f.labelType, normalizeVar)})`
          )
          .join(" ");
        const extension =
          row.extends.node.tag !== "empty-row"
            ? ` ${_printType(row.extends, normalizeVar)}`
            : "";
        return `(cases ${fields}${extension})`;
      }

      case "values":
        return printValuesType(type.node.args[0], normalizeVar, simplify);

      case "->": {
        const { op, args } = type.node;
        return (
          "(" +
          [op, ...args]
            .map((t, i) => _printType(t, normalizeVar, i === args.length))
            .join(" ") +
          ")"
        );
      }

      default: {
        return genericApp(type.node.op, type.node.args);
      }
    }
  } else {
    return genericApp(type.node.op, type.node.args);
  }
}

function _printType(
  type: Type,
  normalizeVar: (t: TVar) => string,
  simplify = false
): string {
  switch (type.node.tag) {
    case "constant":
      return type.node.name;
    case "application":
      return printApplicationType({ node: type.node }, normalizeVar, simplify);
    case "type-variable":
      return normalizeVar({ node: type.node });
    case "empty-row":
    case "row-extension":
      throw new InvariantViolation(`Can't print ${type.node.tag} types`);
  }
}

function printTypeWithOptionalNormalization(rawType: Type, normalize = true) {
  const normalizeVar = normalize
    ? createVariableNormalizer()
    : (t: TVar) => t.node.name;
  return _printType(rawType, normalizeVar);
}

export { printTypeWithOptionalNormalization as printType };
