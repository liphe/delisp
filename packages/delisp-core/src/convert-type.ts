import { capitalize } from "./utils";
import { ConvertError, parseRecord } from "./convert-utils";
import { printHighlightedExpr } from "./error-report";
import { readFromString } from "./reader";
import { last } from "./utils";

import {
  isSymbolOfName,
  ASExpr,
  ASExprList,
  ASExprMap,
  ASExprSymbol,
  ASExprVector
} from "./sexpr";

import {
  emptyRow,
  TConstant,
  TVar,
  Type,
  tApp,
  tBoolean,
  tNumber,
  tRecord,
  tMultiValuedFunction,
  tValues,
  tCases,
  tEffect,
  tString,
  tUserDefined,
  tVar,
  tVector,
  tVoid,
  tNone,
  tcArrow,
  tcStar,
  TypeSchema
} from "./types";

import { TypeWithWildcards } from "./type-wildcards";

/** Return true if a symbol is a valid name for a user defined type, false otherwise. */
export function userDefinedType(expr: ASExprSymbol): boolean {
  return /^[A-Z]/.test(expr.name);
}

/** Check if a symbol is a valid user defined type name or throw a user-friendly error otherwise. */
export function checkUserDefinedTypeName(expr: ASExprSymbol): void {
  if (!userDefinedType(expr)) {
    throw new ConvertError(
      printHighlightedExpr(
        `'${
          expr.name
        }' is not a valid name for a type. Type names should start with a capital letter. Try '${capitalize(
          expr.name
        )}'?`,
        expr.location
      )
    );
  }
}

function convertSymbol(expr: ASExprSymbol): TVar | TConstant {
  switch (expr.name) {
    case "->":
      return tcArrow;
    case "boolean":
      return tBoolean;
    case "number":
      return tNumber;
    case "string":
      return tString;
    case "void":
      return tVoid;
    case "none":
      return tNone;
    case "*":
      return tcStar;
    default:
      return userDefinedType(expr) ? tUserDefined(expr.name) : tVar(expr.name);
  }
}

function convertEffect(effects: ASExpr[]): Type {
  const labels = effects.map(e => {
    if (e.tag !== "symbol") {
      throw new ConvertError(
        printHighlightedExpr(`not a valid effect`, e.location)
      );
    }
    return e.name;
  });
  const last = labels.slice(-2);
  if (last.length === 2 && last[0] === "|") {
    return tEffect(labels.slice(0, -2), tVar(last[1]));
  } else {
    return tEffect(labels);
  }
}

function convertCases(_op: ASExpr, args: ASExpr[]): Type {
  const lastArg = last(args);
  let tail: ASExpr | undefined;

  if (lastArg && lastArg.tag === "symbol" && !lastArg.name.startsWith(":")) {
    tail = args.pop();
  }

  function parseCase(c: ASExpr): { label: string; type: Type } {
    if (c.tag === "symbol" && c.name.startsWith(":")) {
      return { label: c.name, type: tVoid };
    } else if (c.tag === "list") {
      if (c.elements.length !== 2) {
        throw new ConvertError(
          printHighlightedExpr(
            `Invalid case: expected exaclty two elements`,
            c.location
          )
        );
      }
      const [keyword, innerType] = c.elements;
      if (keyword.tag !== "symbol" || !keyword.name.startsWith(":")) {
        throw new ConvertError(
          printHighlightedExpr(
            `Invalid case: expected valid keyword`,
            keyword.location
          )
        );
      }
      return { label: keyword.name, type: convert_(innerType) };
    } else {
      throw new ConvertError(printHighlightedExpr(`Invalid case`, c.location));
    }
  }

  return tCases(args.map(parseCase), tail && convert_(tail));
}

function convertValues(expr: ASExpr): Type[] {
  // Check if X is the symbol VALUES
  if (
    expr.tag === "list" &&
    expr.elements.length > 0 &&
    isSymbolOfName(expr.elements[0], "values")
  ) {
    return expr.elements.slice(1).map(convert_);
  } else {
    return [convert_(expr)];
  }
}

function convertList(expr: ASExprList): Type {
  const [op, ...args] = expr.elements;
  if (op.tag === "symbol" && op.name === "effect") {
    return convertEffect(args);
  } else if (op.tag === "symbol" && op.name === "cases") {
    return convertCases(op, args);
  } else if (op.tag === "symbol" && op.name === "->") {
    if (args.length < 2) {
      throw new ConvertError(`Missing some arguments!`);
    }
    const fnargs = args.slice(0, -2);
    const fneffect = args[args.length - 2];
    const fnout = args[args.length - 1];
    return tMultiValuedFunction(
      fnargs.map(convert_),
      convert_(fneffect),
      tValues(convertValues(fnout))
    );
  } else {
    const opType = convert_(op);
    return tApp(opType, ...args.map(convert_));
  }
}

function convertVector(expr: ASExprVector): Type {
  if (expr.elements.length !== 1) {
    throw new ConvertError(
      printHighlightedExpr("Expected exactly 1 argument", expr.location)
    );
  }
  return tVector(convert_(expr.elements[0]));
}

function convertMap(expr: ASExprMap): Type {
  const { fields, tail } = parseRecord(expr);

  return tRecord(
    fields.map(({ label, value }) => ({
      label: label.name,
      type: convert_(value)
    })),
    tail ? convert_(tail) : emptyRow
  );
}

/* Try to convert a S-Expression into a type. */
function convert_(expr: ASExpr): Type {
  switch (expr.tag) {
    case "list":
      return convertList(expr);
    case "symbol":
      return convertSymbol(expr);
    case "vector":
      return convertVector(expr);
    case "map":
      return convertMap(expr);
    default:
      throw new ConvertError(
        printHighlightedExpr("Not a valid type", expr.location)
      );
  }
}

function convertToTypeWithWildcards(expr: ASExpr): TypeWithWildcards {
  return new TypeWithWildcards(convert_(expr));
}

export { convertToTypeWithWildcards as convert };

export function readType(source: string): TypeSchema {
  const t = convertToTypeWithWildcards(readFromString(source));
  return t.generalize();
}
