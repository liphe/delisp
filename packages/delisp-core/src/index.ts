import { convert as convertSyntax } from "./convert";

import { readFromString } from "./reader";

export { collectConvertErrors } from "./convert";

export {
  compileToString,
  compileModuleToString,
  moduleEnvironment
} from "./compiler";
export { createContext, evaluate, evaluateModule } from "./eval";
export {
  inferType,
  inferModule,
  inferExpressionInModule,
  defaultEnvironment
} from "./infer";
export { Type } from "./types";
export { printType } from "./type-printer";
export {
  isDeclaration,
  isDefinition,
  isExpression,
  isExport,
  isTypeAlias
} from "./syntax";
export { default as primitives } from "./primitives";

export { printHighlightedExpr } from "./error-report";

export function readSyntax(source: string) {
  return convertSyntax(readFromString(source));
}

export { findSyntaxByOffset, findSyntaxByRange } from "./syntax-utils";

export {
  createModule,
  readModule,
  addToModule,
  removeModuleDefinition,
  removeModuleTypeDefinition,
  moduleDefinitions,
  moduleDefinitionByName
} from "./module";

export { generateTSModuleDeclaration } from "./typescript-generation";

export { Encoder } from "./prettier";
export { pprintAs, pprint, pprintModuleAs, pprintModule } from "./printer";
