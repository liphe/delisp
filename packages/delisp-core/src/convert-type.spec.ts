import { readFromString } from "../src/reader";
import { convert as convert_ } from "./convert-type";
import { tFn, tNumber, tString, tVar, Type } from "./types";
import { printType } from "./type-printer";

function readAndConvert(x: string) {
  return convert_(readFromString(x)).noWildcards();
}

function failedType(x: string) {
  let result: Type | string;

  try {
    result = readAndConvert(x);
  } catch (err) {
    result = `\n${err.message}`;
  }

  if (typeof result === "string") {
    return result;
  } else {
    throw new Error(
      `The type is expected to fail, but returned ${printType(result, false)}`
    );
  }
}

describe("convertType", () => {
  it("should convert to numbers", () => {
    expect(readAndConvert("number")).toMatchObject(tNumber);
    expect(readAndConvert("  number  ")).toMatchObject(tNumber);
  });

  it("should convert to strings", () => {
    expect(readAndConvert("string")).toMatchObject(tString);
    expect(readAndConvert("  string  ")).toMatchObject(tString);
  });

  it("should convert to symbols", () => {
    expect(readAndConvert("a")).toMatchObject(tVar("a"));
    expect(readAndConvert("  b  ")).toMatchObject(tVar("b"));
  });

  it("should convert to functions", () => {
    expect(readAndConvert("  (->  string _ number)  ")).toMatchObject(
      tFn([tString], tVar("_"), tNumber)
    );

    expect(readAndConvert("(-> string (-> string _ c) _ c)")).toMatchObject(
      tFn([tString, tFn([tString], tVar("_"), tVar("c"))], tVar("_"), tVar("c"))
    );
  });

  it("should read extensible record", () => {
    expect(readAndConvert("{:x number :y number | a}")).toMatchSnapshot();
  });

  it("should fail for invalid syntax of extensible records", () => {
    expect(() => {
      readAndConvert("{| a :x number}");
    }).toThrow();
  });

  it("should fail for duplicated labels of extensible records", () => {
    expect(() => {
      readAndConvert("{:x 10 :x 20}");
    }).toThrow();
  });

  it("should detect incorrect types", () => {
    expect(failedType("1")).toMatchSnapshot();
    expect(failedType(`"hello"`)).toMatchSnapshot();
    expect(failedType("(1 2 3)")).toMatchSnapshot();
    expect(failedType(`("hello" "world")`)).toMatchSnapshot();
    expect(failedType(`(effect 1`)).toMatchSnapshot();
  });

  describe("Effects", () => {
    it("should convert empty effect", () => {
      expect(readAndConvert("(effect)")).toMatchSnapshot();
    });

    it("should convert simple effect with one labrel", () => {
      expect(readAndConvert("(effect console)")).toMatchSnapshot();
    });

    it("should convert effect with multiple labels", () => {
      expect(readAndConvert("(effect console async)")).toMatchSnapshot();
    });

    it("should convert open effect with multiple labels", () => {
      expect(readAndConvert("(effect console async | a)")).toMatchSnapshot();
    });
  });

  describe("Cases", () => {
    it("should read empty case", () => {
      expect(readAndConvert("(cases)")).toMatchSnapshot();
    });
    it("should read fully polymorphic case", () => {
      expect(readAndConvert("(cases a)")).toMatchSnapshot();
    });
    it("should read basic case", () => {
      expect(
        readAndConvert("(cases (:person string) (:machine number))")
      ).toMatchSnapshot();
    });
    it("should enumeration cases type", () => {
      expect(
        readAndConvert("(cases (:person string) (:machine number) :other)")
      ).toMatchSnapshot();
    });
    it("should detect incorrect cases", () => {
      expect(failedType(`(cases {} {})`)).toMatchSnapshot();
    });
  });

  describe("Values type", () => {
    it("should read the single basic type", () => {
      expect(readAndConvert("(values number)")).toMatchSnapshot();
    });

    it("should read the two values type", () => {
      expect(readAndConvert("(values string number)")).toMatchSnapshot();
    });

    it("should read open value types", () => {
      expect(readAndConvert("(values string | a)")).toMatchSnapshot();
    });
  });
});
