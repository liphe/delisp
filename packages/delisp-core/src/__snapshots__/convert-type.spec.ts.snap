// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertType Cases should detect incorrect cases 1`] = `
"
file:1:8: Invalid case
(cases {} {})
-------^"
`;

exports[`convertType Cases should enumeration cases type 1`] = `
Object {
  "node": Object {
    "args": Array [
      Object {
        "node": Object {
          "extends": Object {
            "node": Object {
              "extends": Object {
                "node": Object {
                  "extends": Object {
                    "node": Object {
                      "tag": "empty-row",
                    },
                  },
                  "label": ":other",
                  "labelType": Object {
                    "node": Object {
                      "name": "void",
                      "tag": "constant",
                    },
                  },
                  "tag": "row-extension",
                },
              },
              "label": ":machine",
              "labelType": Object {
                "node": Object {
                  "name": "number",
                  "tag": "constant",
                },
              },
              "tag": "row-extension",
            },
          },
          "label": ":person",
          "labelType": Object {
            "node": Object {
              "name": "string",
              "tag": "constant",
            },
          },
          "tag": "row-extension",
        },
      },
    ],
    "op": Object {
      "node": Object {
        "name": "cases",
        "tag": "constant",
      },
    },
    "tag": "application",
  },
}
`;

exports[`convertType Cases should read basic case 1`] = `
Object {
  "node": Object {
    "args": Array [
      Object {
        "node": Object {
          "extends": Object {
            "node": Object {
              "extends": Object {
                "node": Object {
                  "tag": "empty-row",
                },
              },
              "label": ":machine",
              "labelType": Object {
                "node": Object {
                  "name": "number",
                  "tag": "constant",
                },
              },
              "tag": "row-extension",
            },
          },
          "label": ":person",
          "labelType": Object {
            "node": Object {
              "name": "string",
              "tag": "constant",
            },
          },
          "tag": "row-extension",
        },
      },
    ],
    "op": Object {
      "node": Object {
        "name": "cases",
        "tag": "constant",
      },
    },
    "tag": "application",
  },
}
`;

exports[`convertType Cases should read empty case 1`] = `
Object {
  "node": Object {
    "args": Array [
      Object {
        "node": Object {
          "tag": "empty-row",
        },
      },
    ],
    "op": Object {
      "node": Object {
        "name": "cases",
        "tag": "constant",
      },
    },
    "tag": "application",
  },
}
`;

exports[`convertType Cases should read fully polymorphic case 1`] = `
Object {
  "node": Object {
    "args": Array [
      Object {
        "node": Object {
          "name": "a",
          "tag": "type-variable",
          "userSpecified": false,
        },
      },
    ],
    "op": Object {
      "node": Object {
        "name": "cases",
        "tag": "constant",
      },
    },
    "tag": "application",
  },
}
`;

exports[`convertType Effects should convert effect with multiple labels 1`] = `
Object {
  "node": Object {
    "args": Array [
      Object {
        "node": Object {
          "extends": Object {
            "node": Object {
              "extends": Object {
                "node": Object {
                  "tag": "empty-row",
                },
              },
              "label": "async",
              "labelType": Object {
                "node": Object {
                  "name": "void",
                  "tag": "constant",
                },
              },
              "tag": "row-extension",
            },
          },
          "label": "console",
          "labelType": Object {
            "node": Object {
              "name": "void",
              "tag": "constant",
            },
          },
          "tag": "row-extension",
        },
      },
    ],
    "op": Object {
      "node": Object {
        "name": "effect",
        "tag": "constant",
      },
    },
    "tag": "application",
  },
}
`;

exports[`convertType Effects should convert empty effect 1`] = `
Object {
  "node": Object {
    "args": Array [
      Object {
        "node": Object {
          "tag": "empty-row",
        },
      },
    ],
    "op": Object {
      "node": Object {
        "name": "effect",
        "tag": "constant",
      },
    },
    "tag": "application",
  },
}
`;

exports[`convertType Effects should convert open effect with multiple labels 1`] = `
Object {
  "node": Object {
    "args": Array [
      Object {
        "node": Object {
          "extends": Object {
            "node": Object {
              "extends": Object {
                "node": Object {
                  "name": "a",
                  "tag": "type-variable",
                  "userSpecified": false,
                },
              },
              "label": "async",
              "labelType": Object {
                "node": Object {
                  "name": "void",
                  "tag": "constant",
                },
              },
              "tag": "row-extension",
            },
          },
          "label": "console",
          "labelType": Object {
            "node": Object {
              "name": "void",
              "tag": "constant",
            },
          },
          "tag": "row-extension",
        },
      },
    ],
    "op": Object {
      "node": Object {
        "name": "effect",
        "tag": "constant",
      },
    },
    "tag": "application",
  },
}
`;

exports[`convertType Effects should convert simple effect with one labrel 1`] = `
Object {
  "node": Object {
    "args": Array [
      Object {
        "node": Object {
          "extends": Object {
            "node": Object {
              "tag": "empty-row",
            },
          },
          "label": "console",
          "labelType": Object {
            "node": Object {
              "name": "void",
              "tag": "constant",
            },
          },
          "tag": "row-extension",
        },
      },
    ],
    "op": Object {
      "node": Object {
        "name": "effect",
        "tag": "constant",
      },
    },
    "tag": "application",
  },
}
`;

exports[`convertType Values type should read open value types 1`] = `
Object {
  "node": Object {
    "args": Array [
      Object {
        "node": Object {
          "extends": Object {
            "node": Object {
              "name": "a",
              "tag": "type-variable",
              "userSpecified": false,
            },
          },
          "label": "0",
          "labelType": Object {
            "node": Object {
              "name": "string",
              "tag": "constant",
            },
          },
          "tag": "row-extension",
        },
      },
    ],
    "op": Object {
      "node": Object {
        "name": "values",
        "tag": "constant",
      },
    },
    "tag": "application",
  },
}
`;

exports[`convertType Values type should read the single basic type 1`] = `
Object {
  "node": Object {
    "args": Array [
      Object {
        "node": Object {
          "extends": Object {
            "node": Object {
              "tag": "empty-row",
            },
          },
          "label": "0",
          "labelType": Object {
            "node": Object {
              "name": "number",
              "tag": "constant",
            },
          },
          "tag": "row-extension",
        },
      },
    ],
    "op": Object {
      "node": Object {
        "name": "values",
        "tag": "constant",
      },
    },
    "tag": "application",
  },
}
`;

exports[`convertType Values type should read the two values type 1`] = `
Object {
  "node": Object {
    "args": Array [
      Object {
        "node": Object {
          "extends": Object {
            "node": Object {
              "extends": Object {
                "node": Object {
                  "tag": "empty-row",
                },
              },
              "label": "1",
              "labelType": Object {
                "node": Object {
                  "name": "number",
                  "tag": "constant",
                },
              },
              "tag": "row-extension",
            },
          },
          "label": "0",
          "labelType": Object {
            "node": Object {
              "name": "string",
              "tag": "constant",
            },
          },
          "tag": "row-extension",
        },
      },
    ],
    "op": Object {
      "node": Object {
        "name": "values",
        "tag": "constant",
      },
    },
    "tag": "application",
  },
}
`;

exports[`convertType should detect incorrect types 1`] = `
"
file:1:1: Not a valid type
1
^"
`;

exports[`convertType should detect incorrect types 2`] = `
"
file:1:1: Not a valid type
\\"hello\\"
^"
`;

exports[`convertType should detect incorrect types 3`] = `
"
file:1:2: Not a valid type
(1 2 3)
-^"
`;

exports[`convertType should detect incorrect types 4`] = `
"
file:1:2: Not a valid type
(\\"hello\\" \\"world\\")
-^"
`;

exports[`convertType should detect incorrect types 5`] = `
"
file:1:10: Expected close parenthesis, number or symbol, double quote, list, open parenthesis, open square bracket, or open curly brace
(effect 1
---------^"
`;

exports[`convertType should read extensible record 1`] = `
Object {
  "node": Object {
    "args": Array [
      Object {
        "node": Object {
          "extends": Object {
            "node": Object {
              "extends": Object {
                "node": Object {
                  "name": "a",
                  "tag": "type-variable",
                  "userSpecified": false,
                },
              },
              "label": ":y",
              "labelType": Object {
                "node": Object {
                  "name": "number",
                  "tag": "constant",
                },
              },
              "tag": "row-extension",
            },
          },
          "label": ":x",
          "labelType": Object {
            "node": Object {
              "name": "number",
              "tag": "constant",
            },
          },
          "tag": "row-extension",
        },
      },
    ],
    "op": Object {
      "node": Object {
        "name": "record",
        "tag": "constant",
      },
    },
    "tag": "application",
  },
}
`;
